File: 0048-rotate-image/0048-rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    // transpose
    for(let i=0;i<matrix.length;i++){
        for(let j=i;j<matrix[i].length;j++){
            let temp= matrix[i][j];
            matrix[i][j]=matrix[j][i];
            matrix[j][i]=temp
        }
    }
    for(let i=0;i<matrix.length;i++){
        let k=0,j=matrix[i].length-1;
        while(k<j){
           let temp= matrix[i][k];
            matrix[i][k]=matrix[i][j];
            matrix[i][j]=temp
            k++;
            j--;
        }
    }
};

File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // transpose
        for(int i=0;i<matrix.size();i++){
            for(int j=i+1;j<matrix.size();j++){
                swap(matrix[i][j],matrix[j][i]);
            }
        }
        // reverse each row
        for(int i=0;i<matrix.size();i++){
            int k=0,j=matrix[0].size()-1;
            while(k<j){
                swap(matrix[i][k],matrix[i][j]);
                k++;
                j--;
            }
        }
    }
};